### ISSUE TITLE
Unicord fails to emulate VMOV on ARM (actually Thumb)

### ISSUE THREAD

**Issue by mkravchik** [unicorn-engine/unicorn#446]
Hi,

I have ARMv7 code which Unicorn fails to emulate. Specifically, the code looks like this in IDA:
__text:00012DB2 C0 EF 10 00                 VMOV.I32        D16, #0 ; Vector Move

Unicorn fails with error:
ERROR: Invalid instruction (UC_ERR_INSN_INVALID)

Are vector instructions supported? What CPU architecture unicorn is based on for ARM? Is there any way to control it?

Thanks!

**Comment by aquynh** [comment]
yes, VMOV should be supported. could you submit a testcase (put it under `tests/unit/` or `tests/regress/`) for this case, so we can confirm the issue & test the fix?

**Comment by mkravchik** [comment]
Wow thanks for such a quick response.
I've never submitted a test case, how do I do that?
The following code fails on emulator but works on a real 32 bit iPhone 5c device

from __future__ import print_function
from unicorn import *
from unicorn.arm_const import *

# memory address where emulation starts
ADDRESS    = 0x10000

code = "\xC0\xEF\x10\x00" #                 VMOV.I32        D16, #0 ; Vector Move
# Initialize emulator in thumb mode
mu = Uc(UC_ARCH_ARM, UC_MODE_THUMB)

# map 2MB memory for this emulation
mu.mem_map(ADDRESS, 2 * 1024 * 1024)

# write machine code to be emulated to memory
mu.mem_write(ADDRESS, code)

# emulate machine code in infinite time
mu.emu_start(ADDRESS, ADDRESS + len(code))

**Comment by aquynh** [comment]
this means you should send a pull request to put your testcase (a Python script) under directory `tests/regress`.

you can take this PR as an example: https://github.com/unicorn-engine/unicorn/pull/434.
it script is at https://github.com/unicorn-engine/unicorn/blob/master/tests/regress/jumping.py

**Comment by mkravchik** [comment]
I just did. #447

**Comment by aquynh** [comment]
anyway, looks like you are missing a step here: you should enable VFP first before running this instruction, no?

according to http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0449b/ch02s01s02.html, we have "... From reset, the VFP extension is disabled. Any attempt to execute a VFP instruction results in an Undefined Instruction exception being taken."

see above link for how to modify your code for VFP, so this sample can work.

**Comment by mkravchik** [comment]
Hi,

Thanks a lot for the tip. However, I can't get the set up to work. According to the link above I should issue instructions:

	MRC p15, 0, r0, c1, c1, 2
	ORR r0, r0, #3<<10 
	MCR p15, 0, r0, c1, c1, 2
	LDR r0, =(0xF << 20)
	MCR p15, 0, r0, c1, c0, 2
	MOV r3, #0x40000000 
	VMSR FPEXC, r3

I compiled them and got the following bytes

00000012	ee110f51	mrc	p15, #0x0, r0, c1, c1, #0x2
00000016	f4406040	orr	r0, r0, #0xc00
0000001a	ee010f51	mcr	p15, #0x0, r0, c1, c1, #0x2
0000001e	    4808	ldr	r0, [pc, #0x20]
00000020	ee010f50	mcr	p15, #0x0, r0, c1, c0, #0x2
00000024	f04f4380	mov.w	r3, #0x40000000
00000028	eee83a10	vmsr	fpexc, r3

or using mrc2
00000012	fe110f51	mrc2	p15, #0x0, r0, c1, c1, #0x2
00000016	f4406040	orr	r0, r0, #0xc00
0000001a	fe010f51	mcr2	p15, #0x0, r0, c1, c1, #0x2
0000001e	    4808	ldr	r0, [pc, #0x20]
00000020	fe010f50	mcr2	p15, #0x0, r0, c1, c0, #0x2
00000024	f04f4380	mov.w	r3, #0x40000000
00000028	eee83a10	vmsr	fpexc, r3

However, trying to run emulation of them fails on the first instruction with the same ERROR: Invalid instruction (UC_ERR_INSN_INVALID)

Do you know what am I doing wrong or point me to a working example?

**Comment by aquynh** [comment]
Hope some ARM experts chime in to help, but I would suggest to make sure your code run on real ARM machine first (or even on Qemu if you dont have the hardware). Only when you confirm that your code is correct, then there is a bug in Unicorn.

**Comment by mkravchik** [comment]
Following your advice I tried to run the code on my ARM7s iPhone 5c. MRC instructions indeed give illegal instruction, but the original one I wanted,
`vmov.i32	d16, #0x0`
runs with no issues without any prior setup.
Please advise. And fails in Unicorn

**Comment by mkravchik** [comment]
I tried to convince emu to allow me to use VFP instructions (if I read correctly MRC with p15 is a privileged instruction - how am I supposed to call these in the emulator?)
So I went to unicorn/target-arm/cpu.c and just set always (regardless to #if defined(CONFIG_USER_ONLY)) 

/* For user mode we must enable access to coprocessors */
env->vfp.xregs[ARM_VFP_FPEXC] = 1 << 30;

I made ./make and then sudo ./make install but saw no changes with regards to the behaviour.
Could anyone guide me on how to proceed? How do I troubleshoot emu?

**Comment by aquynh** [comment]
can you update your testcase at PR #447 with this latest change, so i can merge it?

thanks.

**Comment by mkravchik** [comment]
I've found the proper way to set it up, so I'll close the issue.
In order to enable VFP operations one needs to run the following sequence:
#enable VFP
'''
0000bfae	f44f0070	mov.w	r0, #0xf00000
0000bfb2	ee010f50	mcr	p15, #0x0, r0, c1, c0, #0x2
0000bfb6	f3bf8f6f	isb	sy
0000bfba	f04f4380	mov.w	r3, #0x40000000
0000bfbe	eee83a10	vmsr	fpexc, r3
0000bfc2	efc00010	vmov.i32	d16, #0x0
'''
ENABLE_VFP_CODE = "\x4f\xf4\x70\x00\x01\xee\x50\x0f\xbf\xf3\x6f\x8f\x4f\xf0\x80\x43\xe8\xee\x10\x3a"
What I tried before (and appears in the comments) is for CPUs with TrustZone and it is not supported by Unicorn.

**Comment by aquynh** [comment]
Great, but it is highly possible that others will have this problem again, and come here for similar question. So it is a good idea to update this PR with your new code, i will just merge it.

Thanks.

**Comment by TehRick** [comment]
I wanted this kind of control too, and found this bug. 
After a few hours I found out the cause: JZ/JNZ translation tests cc_dst instead of the carry flag itself.
Here is my own fix in unicorn.c : x86_reg_write : 
```c++
case UC_X86_REG_EFLAGS:
    cpu_load_eflags(&X86_CPU(uc, mycpu)->env, *(uint64_t *)value, -1);
    X86_CPU(uc, mycpu)->env.eflags0 = *(uint64_t *)value | 0x2;
    /* forces Z flag at qemu level (for JZ/JNZ) */
    if (*(uint64_t *)value & CC_Z) {
        X86_CPU(uc, mycpu)->env.cc_dst = 0;
    } else if (X86_CPU(uc, mycpu)->env.cc_dst == 0) {
        X86_CPU(uc, mycpu)->env.cc_dst = 0x0F012CED;
    }
    break;
```
(should be done with uint32_t in mode_32 too)

I didn't want to put it directly in cpu_load_eflags since it's specific to code hooks.
There is another way of doing it, that is to change how JZ is translated.. but it'd be 1 more instruction in the tb for everyone, while we don't often set the Z flag..
I also modified this regress test (jumping.py) to check both flag switch cases (0->1 and 1->0).

Should I propose a pull request ?
My next target is the self-modifying code bug, harder, better, stronger ;)
